docs/source/blogs/tech_blog/blog10_ADP_Balance_Strategy.md

```
@@ -0,0 +1,209 @@
+For example, assume all ranks have contexts of the same length, and each rank has M requests. Let the context phase duration be `time(ctx)` and the generation phase duration be `time(gen)`. Starting from iteration `i`, for N iterations, one new request arrives per iteration. With the original method, one rank processes a context in each iteration, making the per-iteration time `time(ctx)`. The total time would be `time(ctx) * N`. The `balance_ratio` would be `(ctx_len + (M-1) + M * (N - 1)) / (N * ctx_len)`. Here, `gij` represents the j-th generation request on rank i, and `*Cj*` represents the j-th context request.
+
+```text
+iter_i:     [*C0*, g01, ..., g0M], [g10, g11, ..., g1M], ..., [gN0, gN1, ..., gnM]
+iter_i+1:   [g00, g01, ..., g0M], [*C1*, g11, ..., g1M], ..., [gN0, gN1, ..., gnM]
+...
+iter_i+N-1: [g00, g01, ..., g0M], [g10, g11, ..., g1M], ..., [*CN*, gN1, ..., gNM]
+```
+
+If we use our ADP balance method, all ranks can wait for the first N-1 iterations. In the N-th iteration, when the number of context requests equals the number of ranks, they are scheduled together. In this scenario, the execution time for the first N-1 iterations is `time(gen)`, and the last iteration takes `time(ctx)`. The total time is `time(gen) * (N-1) + time(ctx)`. Here, the `balance_ratio=1`, achieving perfect balance in every step. The state of each rank per iteration is as follows:
+
+```
```

markdownlint-cli2 (0.17.2)

Fenced code blocks should have a language specified

(MD040, fenced-code-language)